### PR TITLE
feat: add configurable checksum support for uploads

### PR DESCRIPTION
**Pull request by andrewsg**
Adds configurable checksum support for MultipartUpload and ResumableUpload.

SimpleUpload is not supported because the GCS json API does not accept checksums defined in headers, only in metadata.

MultipartUpload is checked server-side as originally intended for uploads, and a failure will return a 400 error (InvalidResponse exception). ResumableUpload is checked locally as a workaround for having no way to submit the checksum for server-side uploads, as the metadata is transmitted only in the first chunk. A failure there will result in a DataCorruption exception.

### CODE REVIEW

File: google/resumable_media/_helpers.py [hunk 1]
```diff
@@ -164,3 +179,180 @@ def wait_and_retry(func, get_status_code, retry_strategy):
...
+        return "md5Hash"
+    else:
+        return checksum_type
+
+
+def prepare_checksum_digest(digest_bytestring):
+    """Convert a checksum object into a digest encoded for an HTTP header.
+
+    Args:
+        bytes: A checksum digest bytestring.
```
Comment (frankyn) [inline google/resumable_media/_helpers.py:228]: Could you a `Returns:` section instead of a Note:

File: google/resumable_media/_helpers.py [hunk 2]
```diff
@@ -164,3 +179,180 @@ def wait_and_retry(func, get_status_code, retry_strategy):
...
+
+
+def _get_metadata_key(checksum_type):
+    if checksum_type == "md5":
+        return "md5Hash"
+    else:
+        return checksum_type
+
+
+def prepare_checksum_digest(digest_bytestring):
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply google/resumable_media/_helpers.py:224]: Several other helper functions that are similarly independent of internal specifics are public in this file so I decided this one should be intentionally public as well.

File: google/resumable_media/_upload.py [hunk 3]
```diff
@@ -274,11 +290,20 @@ def _prepare_request(self, data, metadata, content_type):
 
         if not isinstance(data, six.binary_type):
             raise TypeError(u"`data` must be bytes, received", type(data))
+
+        checksum_object = _helpers._get_checksum_object(self._checksum_type)
+        if checksum_object:
+            checksum_object.update(data)
+            actual_checksum = _helpers.prepare_checksum_digest(checksum_object.digest())
+            metadata_key = _helpers._get_metadata_key(self._checksum_type)
+            metadata[metadata_key] = actual_checksum
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply google/resumable_media/_upload.py:299]: Overwriting is the intended behavior, so I'll make sure it's tested and documented.

File: google/resumable_media/_helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (frankyn) [reply google/resumable_media/_helpers.py:224]: If that's the case should the file name be renamed to `helpers.py` instead in the case an end-user wants to use them?

[the remark this answers is not in the document]
Comment (andrewsg) [reply google/resumable_media/_helpers.py:224]: I don't think that would be productive as it would commit us to a public interface and there is no expressed customer demand for that, but I wanted to follow the previous convention of not marking functions in this private module further private if they aren't extremely single-use and don't make strong assumptions about internal structure.

File: google/resumable_media/requests/download.py [hunk 4]
```diff
@@ -14,23 +14,14 @@
...
 
-import base64
-import hashlib
-import logging
-
 import urllib3.response
 
 from google.resumable_media import _download
 from google.resumable_media import common
+from google.resumable_media import _helpers as _root_helpers
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply google/resumable_media/requests/download.py:21]: Solving this by refactoring the requests._helpers file to requests._request_helpers; now _helpers won't be ambiguous and it'll be more clear why there are two.

**Comment by andrewsg** [comment]
PTAL (also, this revision unit and system tests for system-native strings (unicode in py3 and bytes in py2) to ensure everything works as users will actually use the software).
